compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEInsertLoopsForVectorization.cpp

```diff
@@ -56,11 +77,41 @@ class AMDAIEInsertLoopsForVectorizationPass
     tileSizes[numIterators - 3] = 0;
     tileSizes[numIterators - 2] = 0;
     tileSizes[numIterators - 1] = 0;
-    auto opts = linalg::LinalgTilingOptions().setTileSizes(tileSizes);
-    auto tiled = linalg::tileLinalgOp(rewriter, genericOp, opts);
-    const auto &loops = tiled.value().loops;
-    assert(!loops.empty() && "expected at least one loop here");
-    rewriter.replaceOp(genericOp, loops[0]->getResult(0));
+    return tileSizes;
+  }
+
+  /// We tile all but the innermost two dimensions currently because they form
+  /// the smallest tiled M x N dimension of the matmul.
+  static std::optional<SmallVector<int64_t>> formTileSizesForElementwise(
+      linalg::GenericOp genericOp) {
+    auto iteratorTypes = genericOp.getIteratorTypesArray();
+    auto numIterators = iteratorTypes.size();
+    assert(numIterators >= 2 && "expected at least 2 iterators here");
+    SmallVector<int64_t> tileSizes(numIterators, 1);
+    tileSizes[numIterators - 2] = 0;
+    tileSizes[numIterators - 1] = 0;
+    return tileSizes;
+  }
+
+  /// Collapse unit dims of the generic op before tiling for vectorization.
+  static void collapseUnitDims(IRRewriter &rewriter,
+                               linalg::GenericOp &genericOp,
+                               bool isTensorType) {
+    linalg::ControlDropUnitDims options;
+    options.rankReductionStrategy =
+        linalg::ControlDropUnitDims::RankReductionStrategy::ExtractInsertSlice;
+    FailureOr<linalg::DropUnitDimsResult> result =
+        linalg::dropUnitDims(rewriter, genericOp, options);
+    if (failed(result)) {
+      return;
+    }
+    if (isTensorType) {
+      rewriter.replaceOp(genericOp, result->replacements);
+    } else {
+      rewriter.eraseOp(genericOp);
+    }
```

Comment:
Because this pass is now made to work on bufferized linalg ops, it won't have a return value, only operands.
Hence I need to remove this too.

Reply:
`tensor`:
```
%a = linalg.generic .... ins(%b : tensor<1x2x3xf32>) outs(%c : tensor<1x2x3xf32>)
```

`bufferized`:
```
linalg.generic .... ins(%b : memref<1x2x3xf32>) outs(%c : memref<1x2x3xf32>)
```

So, if we're forming new a linalg op at the same spot in the IR, we need to replace the old value with the new value which then gets removed by canonicalization. But we don't have the notion of this replacement in case of bufferized inputs, hence we need to remove it explicitly.